File: benchmark-java-node-php-python/golang/monolith/internal/service/bonus_service.go
package service

import (
	"errors"
	"monolith/internal/domain"
	"monolith/internal/dto"
	"monolith/internal/repository"
	"time"

	"gorm.io/gorm"
)

type BonusService interface {
	CreateBonus(dto dto.BonusDTO) (*domain.Bonus, error)
	GetBonus(id string) (*domain.Bonus, error)
}

type bonusService struct {
	bonusRepo  repository.BonusRepository
	clientRepo repository.ClientRepository
}

func NewBonusService(bonusRepo repository.BonusRepository, clientRepo repository.ClientRepository) BonusService {
	return &bonusService{bonusRepo, clientRepo}
}

func (s *bonusService) CreateBonus(dto dto.BonusDTO) (*domain.Bonus, error) {
	client, err := s.clientRepo.FindByID(dto.ClientID)
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, errors.New("client not found")
		}
		return nil, err
	}

	if !client.Active {
		return nil, errors.New("client is inactive")
	}

	finalAmount := dto.Amount
	if finalAmount > 100 {
		finalAmount *= 1.1
	}

	bonus := &domain.Bonus{
		Amount:         finalAmount,
		Description:    "GOLANG - " + dto.Description,
		ClientID:       dto.ClientID,
		CreatedAt:      time.Now(),
		ExpirationDate: time.Now().AddDate(0, 0, 30),
	}

	if err := s.bonusRepo.Create(bonus); err != nil {
		return nil, err
	}

	return bonus, nil
}

func (s *bonusService) GetBonus(id string) (*domain.Bonus, error) {
	bonus, err := s.bonusRepo.FindByID(id)
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, nil // Return nil correctly for "not found" to represent null
		}
		return nil, err
	}
	return bonus, nil
}

File: benchmark-stacks/stacks/apps/java-webflux/monolith/src/main/java/com/benchmark/monolith/service/BonusService.java
package com.benchmark.monolith.service;

import com.benchmark.monolith.domain.Bonus;
import com.benchmark.monolith.dto.BonusDTO;
import com.benchmark.monolith.repository.BonusRepository;
import com.benchmark.monolith.repository.ClientRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class BonusService {
    private final BonusRepository repository;
    private final ClientRepository clientRepository;

    public Mono<Bonus> createBonus(BonusDTO dto) {
        return clientRepository.findById(dto.clientId())
                .switchIfEmpty(Mono.error(new ResponseStatusException(HttpStatus.NOT_FOUND, "Client not found")))
                .flatMap(client -> {
                    if (!client.active()) {
                        return Mono.error(new ResponseStatusException(HttpStatus.BAD_REQUEST, "Client is inactive"));
                    }

                    BigDecimal finalAmount = dto.amount();
                    if (finalAmount.compareTo(new BigDecimal("100")) > 0) {
                        finalAmount = finalAmount.multiply(new BigDecimal("1.1"));
                    }

                    LocalDateTime expirationDate = LocalDateTime.now().plusDays(30);

                    Bonus toSave = new Bonus(
                            null,
                            finalAmount,
                            "JAVAWEBFLUX - " + dto.description(),
                            dto.clientId(),
                            expirationDate,
                            LocalDateTime.now());
                    return repository.save(toSave);
                });
    }

    public Mono<Bonus> getBonus(Long id) {
        return repository.findById(id);
    }

    public Flux<Bonus> getRecents() {
        // Fetch top 100 bonuses ordered by ID ascending
        return repository.findAll()
                .sort(Comparator.comparing(Bonus::id))
                .take(100)
                .collectList()
                .flatMapMany(list -> {
                    // Then sort in memory by createdAt descending to stress memory
                    list.sort(Comparator.comparing(Bonus::createdAt).reversed());
                    List<Bonus> limited = list.stream()
                            .limit(10)
                            .collect(Collectors.toList());
                    return Flux.fromIterable(limited);
                });
    }
}
